src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.cpp

```diff
@@ -0,0 +1,893 @@
+/*******************************************************************\
+
+Module: Dynamic frame condition checking for function and loop contracts.
+
+Author: Qinheping Hu, qinhh@amazon.com
+Author: Remi Delmas delmasrd@amazon.com
+
+Date: March 2023
+
+\*******************************************************************/
+
+#include "dfcc_cfg_info.h"
+
+#include <util/c_types.h>
+#include <util/fresh_symbol.h>
+#include <util/pointer_expr.h>
+
+#include <analyses/local_may_alias.h>
+#include <analyses/natural_loops.h>
+#include <goto-instrument/contracts/utils.h>
+#include <langapi/language_util.h>
+
+#include "dfcc_check_loop_normal_form.h"
+#include "dfcc_infer_loop_assigns.h"
+#include "dfcc_is_cprover_symbol.h"
+#include "dfcc_library.h"
+#include "dfcc_loop_nesting_graph.h"
+#include "dfcc_loop_tags.h"
+#include "dfcc_root_object.h"
+#include "dfcc_utils.h"
+
+/// Extracts the assigns clause expression from the latch condition
+const exprt &get_assigns(const goto_programt::const_targett &latch_target)
+{
+  return static_cast<const exprt &>(
+    latch_target->condition().find(ID_C_spec_assigns));
+}
```

Comment:
1. Yes, it would be a great idea to have such API for instructions.
2. Here we return `exprt` but later use their operands. Yes, we return `exprt::operandst` for consistency.

Reply:
Let's punt on the API for now. It'll probably have to go in `util` after all...